---
title: "R Notebook to estimate the co-accessiblity of scATAC-seq peaks with Cicero"
output: html_notebook
---

```{r}
# load packages and data
library(data.table)
library(Matrix)
library(monocle3)
library(cicero)

tissue="Spleen"
sample_id="62016_P2"

matrix_data     <- fread(paste0("~/Desktop/scATAC-seq/SC_atac_combined.", tissue, "_", sample_id, ".fixed.nodups.peakmatrix.txt"), header = TRUE, sep = "\t")
col_names       <- colnames(read.table(paste0("~/Desktop/scATAC-seq/SC_atac_combined.", tissue, "_", sample_id, ".fixed.nodups.peakmatrix.txt"), header = TRUE, nrows = 1, sep = "\t"))

rownames(matrix_data)<- as.vector(matrix_data[,4])$annot
indata          <- as.matrix(matrix_data[, -c(1:4)])
rownames(indata)<- as.vector(matrix_data[,4])$annot
```

```{r}
# Convert the data into a sparse matrix format.
indata              <- Matrix(indata, sparse = TRUE)

# Binarize the matrix
indata@x[indata@x>0]<- 1

# Format cell info
cellinfo            <- as.data.frame(colnames(indata))
colnames(cellinfo)  <- "cells"
rownames(cellinfo)  <- cellinfo$cells

peakinfo            <- as.data.frame((matrix_data[,1:4]))
colnames(peakinfo)  <- c("chr", "bp1", "bp2", "site_name")
rownames(peakinfo)  <- as.vector(matrix_data[,4])$annot
```


```{r}
# Make CDS
input_cds <- suppressWarnings(new_cell_data_set(indata, cell_metadata = cellinfo, gene_metadata = peakinfo))
input_cds <- monocle3::detect_genes(input_cds)

#Ensure there are no peaks included with zero reads
input_cds <- input_cds[Matrix::rowSums(exprs(input_cds)) != 0, ] 
print(input_cds)

hist(Matrix::colSums(exprs(input_cds)), breaks = 20)

# Filter cells by peak_count
# Please set an appropriate threshold values according to your data 
min_count <- 2000
max_count <- 15000
input_cds <- input_cds[,Matrix::colSums(exprs(input_cds)) >= min_count] 
input_cds <- input_cds[,Matrix::colSums(exprs(input_cds)) <= max_count] 

set.seed(2017)
input_cds <- detect_genes(input_cds)
input_cds <- estimate_size_factors(input_cds)
input_cds <- preprocess_cds(input_cds, method = "LSI")
input_cds <- reduce_dimension(input_cds, reduction_method = 'UMAP', preprocess_method = "LSI")

plot_cells(input_cds)
```

```{r}
umap_coords<- reducedDims(input_cds)$UMAP
cicero_cds <- make_cicero_cds(input_cds, reduced_coordinates = umap_coords)
chromosome_length <- read.table("~/Desktop/scATAC-seq/mm9_chromosome_length.txt")
conns      <- run_cicero(cicero_cds, chromosome_length) # Takes a few minutes to run

# Save results (Optional)
saveRDS(conns, paste0("~/Desktop/scATAC-seq/res_", tissue, "_", sample_id, "_cicero_connections.Rds"))

# Check results
head(conns)

all_peaks  <- row.names(exprs(input_cds))
write.csv(x = all_peaks, file = paste0("~/Desktop/scATAC-seq/res_", tissue, "_", sample_id, "_all_peaks.csv"))
write.csv(x = conns, file = paste0("~/Desktop/scATAC-seq/res_", tissue, "_", sample_id, "_cicero_connections.csv"))
```



