---
title: "R Notebook to estimate the co-accessiblity of scATAC-seq peaks with Cicero"
output: html_notebook
---

```{r}
# load packages and data
library(cicero)
library(data.table)
library(Matrix)
library(proxy)
library(reshape2)
library(umap)

path  <- '~/Desktop/Buenrostro/count_reads_peaks_output/'
files <- list.files(path,pattern = "\\.txt$")
length(files)
```

```{r}
#assuming tab separated values with a header    
datalist  <- lapply(files, function(x)fread(paste0(path,x))$V4) 
#assuming the same header/columns for all files
datafr    <- do.call("cbind", datalist) 
```

```{r}
dim(datafr)
```


```{r}
df_regions      <- read.csv("~/Desktop/Buenrostro/combined.sorted.merged.bed", sep = '\t',header=FALSE,stringsAsFactors=FALSE)
dim(df_regions)
```

```{r}
peaknames       <- paste(df_regions$V1,df_regions$V2,df_regions$V3,sep = "_")
head(peaknames)
head(sapply(strsplit(files,'\\.'),'[', 1))

colnames(datafr)<- sapply(strsplit(files,'\\.'),'[', 1)
rownames(datafr)<- peaknames
datafr[1:5,1:5]
```

```{r}
dim(datafr)
```


```{r}
# Convert the data into a sparse matrix format.
datafr              <- Matrix(datafr, sparse = TRUE)

# Binarize the matrix
datafr@x[datafr@x>0]<- 1

# Format cell info
cellinfo            <- as.data.frame(colnames(datafr))
colnames(cellinfo)  <- "cells"
rownames(cellinfo)  <- cellinfo$cells

peakinfo            <- as.data.frame(cbind(df_regions, peaknames))
colnames(peakinfo)  <- c("chr", "bp1", "bp2", "site_name")
rownames(peakinfo)  <- peaknames

print(datafr[1:10,1:10])
print(cellinfo[1:10, ])
print(peakinfo[1:10, ])
```


```{r}
# Make CDS
input_cds <- suppressWarnings(new_cell_data_set(datafr, cell_metadata = cellinfo, gene_metadata = peakinfo))
input_cds <- monocle3::detect_genes(input_cds)

#Ensure there are no peaks included with zero reads
input_cds <- input_cds[Matrix::rowSums(exprs(input_cds)) != 0, ] 
print(input_cds)

hist(Matrix::colSums(exprs(input_cds)), breaks = 20)

# Filter cells by peak_count
# Please set an appropriate threshold values according to your data 
min_count <- 2000
max_count <- 15000
input_cds <- input_cds[,Matrix::colSums(exprs(input_cds)) >= min_count] 
input_cds <- input_cds[,Matrix::colSums(exprs(input_cds)) <= max_count] 

set.seed(2017)
input_cds <- detect_genes(input_cds)
input_cds <- estimate_size_factors(input_cds)
input_cds <- preprocess_cds(input_cds, method = "LSI")
input_cds <- reduce_dimension(input_cds, reduction_method = 'UMAP', preprocess_method = "LSI")

plot_cells(input_cds)
```


```{r}
umap_coords<- reducedDims(input_cds)$UMAP
cicero_cds <- make_cicero_cds(input_cds, reduced_coordinates = umap_coords)
chromosome_length <- read.table("~/Desktop/Buenrostro/hg19.chrom.sizes")
conns      <- run_cicero(cicero_cds, chromosome_length) # Takes a few minutes to run

# Save results (Optional)
saveRDS(conns, paste0("~/Desktop/Buenrostro/res_Buenrostro2018_cicero_connections.Rds"))

# Check results
head(conns)

all_peaks  <- row.names(exprs(input_cds))
write.csv(x = all_peaks, file = paste0("~/Desktop/Buenrostro/res_Buenrostro2018_all_peaks.csv"))
write.csv(x = conns, file = paste0("~/Desktop/Buenrostro/res_Buenrostro2018_cicero_connections.csv"))
```



















